Changes in module 'nf-core/bowtie2/align'
--- modules/nf-core/bowtie2/align/main.nf
+++ modules/nf-core/bowtie2/align/main.nf
@@ -1,6 +1,6 @@
 process BOWTIE2_ALIGN {
     tag "$meta.id"
-    label "process_high"
+    label "process_low" // process_high
 
     conda "bioconda::bowtie2=2.4.4 bioconda::samtools=1.16.1 conda-forge::pigz=2.6"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
@@ -9,7 +9,7 @@
 
     input:
     tuple val(meta) , path(reads)
-    tuple val(meta2), path(index)
+    path (index)
     val   save_unaligned
     val   sort_bam
 

************************************************************
